### ISSUE TITLE
Array-typed const should be detoasted during transformation back to an ArrayExpr

### ISSUE THREAD

**Issue by d** [greenplum-db/gpdb#5373]
## TL;DR
We came across this issue when @l-wang, @melanieplageman, and I were working on the iteration that merges Postgres 9.2 into Greenplum. Chasing down the rabbit hole, we found two issues, one of them being the way `pg_locks` is implemented, and the other being how we fold constants in the query before sending the query to ORCA. This is that second issue. (guess this isn't short after all)

### Greenplum version or build
master (6.0 after merging 9.1 but before merging Postgres 9.2, commit 04e39f4a3c as of writing)

### OS version and uname -a
Ubuntu, but this isn't relevant

### autoconf options used ( config.status --config )
Again, not relevant, but here it is:
`./configure --enable-depend --enable-debug --enable-cassert --enable-orca --disable-gpcloud --with-libxml --with-perl --with-python --with-includes=$HOME/workspace/install.orca/include --with-libs=$HOME/workspace/install.orca/lib CFLAGS='-O0' CC='ccache clang-6.0' CXX='ccache clang++-6.0' LDFLAGS='-fuse-ld=gold'`

### Expected behavior
That Greenplum does not crash

### Actual behavior

```
csteam@station2:~/workspace/gpdb$ psql -a -f /tmp/array.sql 
set optimizer_trace_fallback to on;
SET
CREATE OR REPLACE FUNCTION boom()
RETURNS bool AS $$
DECLARE
	mel bool;
	sesh int[];
BEGIN
	sesh := '{42,1}'::int[];
	EXECUTE 'select c = ANY ($1) from (values (42), (0)) yolo(c)' INTO mel USING sesh;
	return mel;
END
$$
LANGUAGE plpgsql VOLATILE;
CREATE FUNCTION
SELECT boom();
psql:/tmp/array.sql:15: INFO:  GPORCA failed to produce a plan, falling back to planner
psql:/tmp/array.sql:15: INFO:  GPORCA failed to produce a plan, falling back to planner
CONTEXT:  SQL statement "SELECT sesh"
PL/pgSQL function "boom" line 7 at EXECUTE statement
psql:/tmp/array.sql:15: FATAL:  Unexpected internal error (arrayfuncs.c:3076)
DETAIL:  FailedAssertion("!(((array)->elemtype) == elmtype)", File: "arrayfuncs.c", Line: 3076)
HINT:  Process 8182 will wait for gp_debug_linger=120 seconds before termination.
Note that its locks and other resources will not be released until then.
CONTEXT:  SQL statement "select c = ANY ($1) from (values (42), (0)) yolo(c)"
PL/pgSQL function "boom" line 7 at EXECUTE statement
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
psql:/tmp/array.sql:15: connection to server was lost

```

### Step to reproduce the behavior

```SQL

CREATE OR REPLACE FUNCTION boom()
RETURNS bool AS $$
DECLARE
	mel bool;
	sesh int[];
BEGIN
	sesh := '{42,1}'::int[];
	EXECUTE 'select c = ANY ($1) from (values (42), (0)) yolo(c)' INTO mel USING sesh;
	return mel;
END
$$
LANGUAGE plpgsql VOLATILE;

SELECT boom();

```

**Comment by d** [comment]
@melanieplageman look at the mad SQL!
